menpo3d/barycentric.py

```diff
@@ -13,12 +13,29 @@ def barycentric_coordinates_for_indices(mesh, tri_index, point):
     return barycentric_coordinates(point, a, b, c)
 
 
-def barycentric_points_from_contained_points(self, pointcloud, indices):
-    bcoords = np.array([barycentric_coordinates_for_indices(self, i, p)
-                        for i, p in zip(indices, pointcloud.points)])
-    # compute the third (degenerate) coordinate as well
-    bcoords_full = np.hstack((bcoords, (1 - bcoords.sum(axis=1))[:, None]))
-    return bcoords_full
+def barycentric_points_from_contained_points(self, pointcloud, tri_index):
+    # http://gamedev.stackexchange.com/questions/23743/whats-the-most-efficient-way-to-find-barycentric-coordinates
+    abc_per_tri = self.points[self.trilist[tri_index]]
+
+    a = abc_per_tri[:, 0, :]
+    b = abc_per_tri[:, 1, :]
+    c = abc_per_tri[:, 2, :]
+    p = pointcloud.points
+
+    v0 = b - a
+    v1 = c - a
+    v2 = p - a
+    d00 = (v0 * v0).sum(axis=1)
+    d01 = (v0 * v1).sum(axis=1)
+    d11 = (v1 * v1).sum(axis=1)
+    d20 = (v2 * v0).sum(axis=1)
+    d21 = (v2 * v1).sum(axis=1)
+    denom = d00 * d11 - d01 * d01
+    v = (d11 * d20 - d01 * d21) / denom
+    w = (d00 * d21 - d01 * d20) / denom
```

Comment:
@patricksnape yes, but I believe only in the degenerate case of a triangle with zero area. Not sure if it's worth guarding against..I guess we could do a try/catch and warn?